klippy/extras/bed_mesh.py

```diff
@@ -1134,6 +1135,13 @@ def initialize(self):
         self._check_incompatible_profiles()
         if "default" in self.profiles:
             self.load_profile("default")
+    def get_profiles(self):
+        result = dict()
+        for name in self.profiles:
+            profile = dict(self.profiles[name])
+            profile["mesh_params"] = dict(profile["mesh_params"])
+            result[name] = profile
+        return result
```

Comment:
Hi @Arksine , thanks for the review.

I understood from the documentation that the items being returned as status should not be allowed to change under the hood.  The profiles can be updated by the other commands to load/save a profile and hence returning a reference to the existing profiles would be subject to such change.  That was one reason for doing it this way.

If I recall correctly, there was also the problem of some of the items (mesh_params) being ordered dictionaries rather than plain ones and that was throwing of things when then trying to store the information from the status into a saved variable or passing it to some other gcode.  I'm not sufficiently knowledgeable on Klipper and Python yet to fully grasp how the passing of such a status dictionary to a gcode parameter works, but it looked to me like a kind of implicit string conversion, which for an OrderedDict ended up being a string that would not get processed correctly by ast.literal_eval.  So, the conversion here was also to get rid of the OrderedDict.  I don't know why the original code was using an OrderedDict for mesh_params.  An alternative could be to remove that and if I would then ignore the rule that returned status should not get changed after returning, then not doing a copy might work ...

Is there any way for modules to provide data back to a macro other than via the get_status()?  If so, then that might be a more appropriate way to allow access to the actual list of profiles.  As said, I might just not be aware of the existence of such a mechanism, hence my current implementation via the get_status().

Please advice how to proceed.

Reply:
While it is true that none of the items referenced in `get_status()` should be modified, I don't believe we take any action to prevent it from happening.  As mentioned above, the only way they could be modified by a template is if the user directly calls a dict method on the result.  Anyone knowledgeable enough to do that in Jinja2 will likely know the source of any pain this induces.

An OrderDict was used to keep the output consistent when saving a mesh to the config file.  To be honest, I may have introduced a regression here in a later commit, as it appears that the mesh params are converted to a regular dict in the `finalize()` method.  This isn't a deal breaker, we could probably remove the ordered dict and just sort the keys prior to setting the new config.

You are correct that the ordered dict would break ast, and the best way of handling it is to just remove it and sort the mesh params in `save_profile()`.  That said, there are also the alternatives methods of saving variables that I outlined below.